Clean up clippy allows

ValuedMammal opened the pull request:
closes #1127 

There are several instances in the code where we allow clippy lints that would otherwise be flagged during regular checks. It would be preferable to minimize the number of "clippy allow" attributes either by fixing the affected areas or setting a specific configuration in `clippy.toml`. In cases where we have to allow a particular lint, it should be documented why the lint doesn't apply.

For context see https://github.com/bitcoindevkit/bdk/issues/1127#issuecomment-1784256647 as well as the commit message details.

One area I'm unsure of is whether `Box`ing a large error in 4fc2216 is the right approach. Logically it makes sense to avoid allocating a needlessly heavy `Result`, but I haven't studied the implications or tradeoffs of such a change.

### Checklists

#### All Submissions:

* [x] I've signed all my commits
* [x] I followed the [contribution guidelines](https://github.com/bitcoindevkit/bdk/blob/master/CONTRIBUTING.md)
* [x] I ran `cargo fmt` and `cargo clippy` before committing

ValuedMammal committed:
refactor: remove old clippy allow attributes

These lints either resolved themselves, or the code has changed such that
they no longer apply, hence they can be removed with no further changes.

`clippy::derivable_impls`
`clippy::needless_collect`
`clippy::almost_swapped`

ValuedMammal committed:
ref(example_cli): add typedefs to reduce type complexity

- Add typedefs to model the result of functions `planned_utxos`
and `init`

- Add new struct `CreateTxChange` to hold any change info
resulting from `create_tx`

These changes help resolve clippy::type_complexity

ValuedMammal committed:
ref(psbt): refactor body of `get_utxo_for` to address `clippy::manual_map`

ValuedMammal committed:
ref(esplora): `Box` a large `esplora_client::Error`

to address `clippy::result_large_err`. Clippy's default large-error-
threshold is 128. `esplora_client::Error` currently has size 272.

ValuedMammal commented:
The first attempt failed CI due to `clippy::iter_nth_zero`. Of course, there's no reason to think the lint could be disregarded in the first place.. I believe I used the wrong invocation of `cargo clippy` on the command line, allowing it to slip by. I suppose the nix experience should eventually help prevent silly mistakes.

After taking another look at spk_iter.rs, I think it's fine to allow the exception (as it was originally), since we're specifically testing `SpkIterator`'s custom implementation of `nth`. An alternative would be to write the tests in a way that doesn't require calling `nth(0)`, but for now this seems like a trivial case not worth worrying about.

LLFourn commented on nursery/tmp_plan/src/requirements.rs:
```diff
@@ -117,6 +117,7 @@ impl From<bip32::Error> for SigningError {
 impl std::error::Error for SigningError {}
 
 impl RequiredSignatures<DescriptorPublicKey> {
+    #[allow(clippy::too_many_arguments)]
```
feel free to just #[allow(clippy::all)]  in this module's lib.rs

danielabrozzoni commented on example-crates/example_cli/src/lib.rs:
```diff
@@ -183,18 +183,17 @@ impl core::fmt::Display for Keychain {
 
-#[allow(clippy::type_complexity)]
+// Alias changeset info in `create_tx`
+pub type ChangeSet = (keychain::ChangeSet<Keychain>, (Keychain, u32));
```
Instead of defining this alias, can we change the return parameter of create_tx to return a structure? It can contain:
- transaction
- keychain_changeset
- change_keychain
- change_index

I think this would make the code a bit clearer, what do you think?

ValuedMammal replied:
Yes, I like that. However, since three of these fields are either all `Some` or all `None`, it might make sense to have a struct just for `change_info` and continue to return a tuple of `(transaction, change_info)` from `create_tx`

For example, here I introduce `CreateTxChange` and make its members match the names of the existing logic in fn `handle_commands`, does that make sense?

```
--- a/example-crates/example_cli/src/lib.rs
+++ b/example-crates/example_cli/src/lib.rs
@@ -181,21 +181,24 @@ impl core::fmt::Display for Keychain {
             Keychain::Internal => write!(f, "internal"),
         }
     }
 }
 
-// Alias changeset info in `create_tx`
-pub type ChangeSet = (keychain::ChangeSet<Keychain>, (Keychain, u32));
+pub struct CreateTxChange {
+    pub index_changeset: keychain::ChangeSet<Keychain>,
+    pub change_keychain: Keychain,
+    pub index: u32,
+}
 
 pub fn create_tx<A: Anchor, O: ChainOracle>(
     graph: &mut KeychainTxGraph<A>,
     chain: &O,
     keymap: &HashMap<DescriptorPublicKey, DescriptorSecretKey>,
     cs_algorithm: CoinSelectionAlgo,
     address: Address,
     value: u64,
-) -> anyhow::Result<(Transaction, Option<ChangeSet>)>
+) -> anyhow::Result<(Transaction, Option<CreateTxChange>)>
 where
     O::Error: std::error::Error + Send + Sync + 'static,
 {
     let mut changeset = keychain::ChangeSet::default();
 
@@ -385,11 +388,15 @@ where
             }
         }
     }
 
     let change_info = if selection_meta.drain_value.is_some() {
-        Some((changeset, (internal_keychain, change_index)))
+        Some(CreateTxChange {
+            index_changeset: changeset,
+            change_keychain: internal_keychain,
+            index: change_index,
+        })
     } else {
         None
     };
 
     Ok((transaction, change_info))
@@ -593,11 +600,16 @@ where
                 let graph = &mut *graph.lock().unwrap();
                 // take mutable ref to construct tx -- it is only open for a short time while building it.
                 let (tx, change_info) =
                     create_tx(graph, chain, keymap, coin_select, address, value)?;
 
-                if let Some((index_changeset, (change_keychain, index))) = change_info {
+                if let Some(CreateTxChange {
+                    index_changeset,
+                    change_keychain,
+                    index,
+                }) = change_info
+                {
                     // We must first persist to disk the fact that we've got a new address from the
                     // change keychain so future scans will find the tx we're about to broadcast.
                     // If we're unable to persist this, then we don't want to broadcast.
                     {
                         let db = &mut *db.lock().unwrap();
```

danielabrozzoni replied:
Yes, sounds good!

danielabrozzoni commented on crates/esplora/src/async_ext.rs:
```diff
@@ -6,11 +6,17 @@ use bdk_chain::{
+type Error = Box<esplora_client::Error>;
+
+/// Alias `Result` to use our `Error`
+type Result<T, E = Error> = core::result::Result<T, E>;
```
Why is this line needed?

ValuedMammal replied:
You're right it's not needed, let's get rid of it. The thought was that it allows omitting the `Err` variant in a function signature similar to what you get from `anyhow::Result`, but I neglected to actually remove them. In this case, it's probably better to be explicit and keep `Error` in the fn signatures.

danielabrozzoni reviewed (commented):
Overall it looks really good, thanks a lot!
A few questions/suggestions:

danielabrozzoni commented on crates/bdk/src/wallet/coin_selection.rs:
```diff
@@ -531,10 +517,19 @@ impl BranchAndBoundCoinSelection {
         optional_utxos.sort_unstable_by_key(|a| a.effective_value);
         optional_utxos.reverse();
 
+        let mut curr_value = required_utxos
```
I don't think it's worth the hassle to refactor the cs module, since it's going away soon. Also see https://github.com/bitcoindevkit/bdk/issues/1127#issuecomment-1784313256

ValuedMammal committed:
ref(wallet): `Wallet::preselect_utxos` now accepts a `&TxParams`

to reduce the number of required function args in order to satisfy
`clippy::too_many_arguments`

ValuedMammal commented:
rebasing now. dropped all the changes to `coin_selection.rs`, and made `ref(tmp_plan)` the last commit

ValuedMammal commented on crates/bdk/src/wallet/mod.rs:
I guess I'm on the fence regarding `preselect_utxos`. @LLFourn would you prefer we left this untouched so it's clear at the call site exactly what information we're passing?

ValuedMammal commented:
Reminder to bump msrv in clippy.toml
```
warning: the MSRV in `clippy.toml` and `Cargo.toml` differ; using `1.57.0` from `clippy.toml`
```

ValuedMammal commented:
Changes in most recent push:

- re-refactored `Psbt::get_utxo_for` to use a match block. I think it conveys the same thing in a more concise way. [compare](https://github.com/bitcoindevkit/bdk/compare/6cdc7f1a95f2a26f65b87ead00cc57b328f18f51..6107725756493ff5f0afa94c25c11d0a598933a1#diff-504dcd23159cb3b14ff38d9c2148749e8a213d292c346c55035f36c0aae7ccd6)

LLFourn commented on nursery/tmp_plan/src/lib.rs:
```diff
@@ -12,7 +12,7 @@
 //! will actually need as well as what locktime or sequence number you need to set.
 //!
-//! Once you've obstained signatures, hash pre-images etc required by the plan, it can create a
+//! Once you've obtained signatures, hash pre-images etc required by the plan, it can create a
```
I'd prefer we didn't touch this crate and just allowed all.

LLFourn commented:
Please revert all changes to the tmp plan crate and just blanket allow everything. Other than that this looks good to go.

ValuedMammal commented:
@LLFourn appreciate your input, thanks for clarifying.

LLFourn commented on example-crates/example_cli/src/lib.rs:
```diff
@@ -643,17 +651,19 @@ where
-    db_default_path: &str,
-) -> anyhow::Result<(
+// Alias the `Result` of `init`
+pub type InitialState<CS, S, C> = (
```
This could actually be made into a struct where each field had a nice comment if you felt up to it.

ValuedMammal replied:
Would you rename this?

LLFourn reviewed (approved):
ACK c4f121109198df3a3f96b507fe76a7f14146a135

Thanks @ValuedMammal!

ValuedMammal commented:
Is anyone opposed to making the type `InitialState` in `example_cli` into a struct with comments?

```rust
/// The initial state returned from [`init`].
pub struct Init<CS: clap::Subcommand, S: clap::Args, C> {
    /// Arguments parsed by the cli.
    pub args: Args<CS, S>,
    /// Descriptor keymap.
    pub keymap: KeyMap,
    /// Keychain-txout index.
    pub index: KeychainTxOutIndex<Keychain>,
    /// Persistence backend.
    pub db: Mutex<Database<C>>,
    /// Initial changeset.
    pub init_changeset: C,
}
```

danielabrozzoni commented:
@ValuedMammal go for it!

ValuedMammal committed:
ref(example_cli): Add new struct Init

for holding the items returned from `example_cli::init`

evanlinjin commented on crates/bdk/src/psbt/mod.rs:
```diff
@@ -35,24 +35,20 @@ pub trait PsbtUtils {
+        let input = match self.inputs.get(input_index) {
+            Some(input) => input,
+            None => return None,
+        };
```
Nit:

```suggestion
        let input = self.inputs.get(input_index)?;
```

ValuedMammal replied:
indeed, this is better.

evanlinjin commented on example-crates/example_cli/src/lib.rs:
```diff
@@ -388,43 +388,46 @@ where
 
-#[allow(clippy::type_complexity)]
+// Alias the elements of `Result` of `planned_utxos`
+pub type Plan<K, A> = (bdk_tmp_plan::Plan<K>, FullTxOut<A>);
```
Nit: can we rename this to `PlannedUTXO`? Thanks!

ValuedMammal replied:
I hope you don't mind I used `PlannedUtxo`.

evanlinjin replied:
Woops sorry I was in Golang world these two weeks.

evanlinjin reviewed (commented):
Looking good! Just a couple of NITs

ValuedMammal commented:
- Rebased onto f099b42005b5310cee51fa2eaa29f94d76753bd7
- Addressed comments by @evanlinjin 

Thank you to all the reviewers! @danielabrozzoni @LLFourn

evanlinjin reviewed (approved):
ACK 1c15cb2f9169bb08a4128c183eeeca80fff97ab7
